libs/vm-modules/tests/unit/ml/ml_model_tests.cpp

```diff
@@ -1579,4 +1591,88 @@ TEST_F(VMModelTests, model_fit_and_refit)
   ASSERT_TRUE(toolkit.Run());
 }
 
+TEST_F(VMModelTests, model_add_input_layer_as_second)
+{
+  static char const *SRC = R"(
+      function main()
+        var data_shape = Array<UInt64>(2);
+        data_shape[0] = 10u64;
+        data_shape[1] = 250u64;
+
+        var model = Model("sequential");
+
+        model.add("dense", 10u64, 10u64);
+        model.addExperimental("input", data_shape);
+      endfunction
+    )";
+
+  ASSERT_TRUE(toolkit.Compile(SRC));
+  ASSERT_FALSE(toolkit.Run(nullptr, ChargeAmount{0}));
+}
+
+TEST_F(VMModelTests, model_add_dense_auto_inputs)
+{
+  static char const *SRC = R"(
+      function main()
+        var data_shape = Array<UInt64>(4);
+        data_shape[0] = 3u64;
+        data_shape[1] = 3u64;
+        data_shape[2] = 3u64;
+        data_shape[3] = 25u64;
+
+        var model = Model("sequential");
+
+        model.addExperimental("input", data_shape);
+        model.addExperimental("dense", 10u64);
+        model.addExperimental("dense", 6u64);
+        model.addExperimental("dense", 1u64);
+        model.compile("scel", "adam");
+
+        var data = Tensor(data_shape);
+        data.fillRandom();
+        model.predict(data);
+      endfunction
+    )";
+
+  ASSERT_TRUE(toolkit.Compile(SRC));
+  ASSERT_TRUE(toolkit.Run(nullptr, ChargeAmount{0}));
+}
+
+TEST_F(VMModelTests, model_add_mixed_auto_inputs)
+{
+  static char const *SRC = R"(
+      function main()
+         var num_channels = 3u64;
+         var image_height = 20u64;
+         var image_width = 40u64;
+
+         var data_shape = Array<UInt64>(4);
+         data_shape[0] = num_channels;
+         data_shape[1] = image_height;
+         data_shape[2] = image_width;
+         data_shape[3] = 25u64;
+
+         var model = Model("sequential");
+
+         model.addExperimental("input", data_shape);
```

Comment:
so the input layer is working; but we don't yet have the right bindings for dense and conv to allow just output shape specification?
But presumably all the underlying code is there for that, we just need to add a couple of functions to vmmodules sequential, right?

Reply:
model.addExperimental("dense", 6u64);  -- this is already here for output-only Dense declaration; regarding Conv2D - it has only "input channels" parameter, it might be auto-deduced from some input shape of previous layer, but I am not ready to tell if this implementation is trivial.

Reply:
Well, the first leyer in a Model could be also Dropout, Flatten, Reshape or another one, so it moght be a good idea to create interface for adding any type of layer to a Model with ptoviding input shape array as a parameter (as it is possible in Keras), however, it is a large change so I would prefer to keep only Dense as an example and an InputLayer as a default working solution for now.